lib/client/ipam.go

```diff
@@ -419,47 +423,58 @@ func (c ipams) assignFromExistingBlock(
 	return ips, nil
 }
 
-// ClaimAffinity claims affinity to the given host for all blocks
+// ClaimAffinity makes a best effort to claim affinity to the given host for all blocks
 // within the given CIDR.  The given CIDR must fall within a configured
-// pool.
-// TODO: Return indicators of claimed vs already claimed by another host.
-func (c ipams) ClaimAffinity(cidr common.IPNet, host *string) error {
+// pool.  Returns a list of blocks that were claimed, as well as a
+// list of blocks that were claimed by another host.
+func (c ipams) ClaimAffinity(cidr common.IPNet, host *string) ([]common.IPNet, []common.IPNet, error) {
 	// Validate that the given CIDR is at least as big as a block.
 	if !largerThanBlock(cidr) {
 		estr := fmt.Sprintf("The requested CIDR (%s) is smaller than the minimum.", cidr.String())
-		return invalidSizeError(estr)
+		return nil, nil, invalidSizeError(estr)
 	}
 
 	// Determine the hostname to use.
 	hostname := decideHostname(host)
+	failed := []common.IPNet{}
+	claimed := []common.IPNet{}
 
 	// Verify the requested CIDR falls within a configured pool.
 	if !c.blockReaderWriter.withinConfiguredPools(common.IP{cidr.IP}) {
 		estr := fmt.Sprintf("The requested CIDR (%s) is not within any configured pools.", cidr.String())
```

Comment:
Do these errors bubble up to the top?  Reason I ask is that i'm not a big fan of an error where the only details about the error are in a description string.  Suppose (much further down the line) we wanted to provide multi-language support, this doesn't lend itself well to that.  Not suggesting you fix this as I'm still not 100% sure what our error handling plan is.